lib/events/roomcanonicalaliasevent.h

```diff
@@ -0,0 +1,86 @@
+// Contains code for RoomCanonicalAlias Event
+#pragma once
+
+#include "stateevent.h"
+
+namespace Quotient {
+namespace EventContent{
+    template <typename T1, typename T2>
+    class SimpleDualContent {
+    public:
+        using first_value_type = T1;
+        using second_value_type = T2;
+
+        template <typename TT1, typename TT2>
+        SimpleDualContent(QString Key1Name, TT1&& value1, QString Key2Name,
+                          TT2&& value2)
+            : value1(std::forward<TT1>(value1))
+            , value2(std::forward<TT2>(value2))
+            , key1(std::move(Key1Name))
+            , key2(std::move(Key2Name))
+        { }
+
+        SimpleDualContent(const QJsonObject& json, QString key1Name,
+                          QString key2Name)
+            : value1(fromJson<T1>(json[key1Name]))
+            , value2(fromJson<T2>(json[key2Name]))
+            , key1(std::move(key1Name))
+            , key2(std::move(key2Name))
+        { }
+
+        QJsonObject toJson() const
+        {
+            return { { key1, Quotient::toJson(value1) },
+                     { key2, Quotient::toJson(value2) } };
+        }
+
+    public:
+        T1 value1;
+        T2 value2;
+
+    protected:
+        QString key1;
+        QString key2;
+    };
+} // namespace EventContent
+
+class RoomCanonicalAliasEvent
+    : public StateEvent<EventContent::SimpleDualContent<QString, QStringList>> {
+public:
+    DEFINE_EVENT_TYPEID("m.room.canonical_alias", RoomCanonicalAliasEvent)
+
+    explicit RoomCanonicalAliasEvent(const QJsonObject& obj)
+        : StateEvent(typeId(), obj, QStringLiteral("alias"),
+                     QStringLiteral("alt_aliases"))
+    { }
+
+    RoomCanonicalAliasEvent(const QString& server, const QString& alias,
+                            const QStringList& alt_aliases)
+        : StateEvent(typeId(), matrixTypeId(), server, QStringLiteral("alias"),
+                     alias, QStringLiteral("alt_aliases"), alt_aliases)
+    { }
+
+    // For compatibility used at Room::setCanonicalAlias
+    explicit RoomCanonicalAliasEvent(const QString& value1)
```

Comment:
Please don't use boilerplate names for parameters you intend for usage by client writers (like this one). And besides, this constructor is not a good idea in itself - see my comment in `room.cpp` on why you do _not_ want compatibility with the current code of `Room::setCanonicalAlias`.